### PR TITLE
Fixing issue with EarlyStopping not working after CustomCallback

### LINKED ISSUE #37587: EarlyStopping Callback not working with Multiple Callbacks
**System information** 
- Have I written custom code  
- OS Platform and Distribution: *Google Colab*
- TensorFlow version: *2.1.0*
- Keras: *2.2.4-tf*
- Python version: *3* 
- CUDA/cuDNN version: No GPU but happens also with it

**Describe the current behavior**
When a CustomCallback  uses `model.predict` or `model.evaluate` as part of `on_epoch_end`,  to for example, track a custom evaluation metric on an test dataset, `model.stop_training` is reset to False even if it was previously set to True by Early Stopping. Training does not stop. Current workaround is putting EarlyStopping as the last callback of the list.

**Describe the expected behavior**
Regardless the order in which callbacks are called, if one of the sets the model to stop training it should stop even if a later callback resets the stop flag.

**Standalone code to reproduce the issue** 
Check this colab with the full example: https://colab.research.google.com/drive/1lw943Ggwkp_wvGxVX-5XqaEJ5qXmrHlA

In short:
```
class MyCallback(keras.callbacks.Callback):
    def __init__(self, test_data):
        super(MyCallback, self).__init__()
        self.test_data = test_data

    def on_epoch_end(self, epoch, logs=None):
        print(f"\n--------- pre-predict stop_training={self.model.stop_training}\n")
        #The problem is in the prediction: if commented ES works fine
        predictions = self.model.predict(self.test_data.batch(512))
        print(f"\n--------- post-predict stop_training={self.model.stop_training}\n")
```
```
es = keras.callbacks.EarlyStopping(patience=2)
myc = MyCallback(test_data)

#This causes EarlyStop not to stop
my_callbacks = [es, myc]
#Either of these works fine
#my_callbacks = [myc, es]
#my_callbacks = [es]
...
model.fit(train_data.batch(512),
          validation_data=validation_data.batch(512),
          epochs=100,
          callbacks=my_callbacks,
          verbose=1)
```

This issue is refered also in the keras repository: https://github.com/keras-team/keras/issues/13381

Some thoughts: I'm not sure if the correct solution would be to discourage the use of predict or evaluate inside a training loop since there may be some other side effects of running one of those to the model.

Anyway, I'm opening the issue and submitting a pull request to fix this.

### PR DESCRIPTION
**Pull request by eduardofv**
When a CustomCallback uses internally model.predict or model.evaluate as part of on_epoch_end, for instance, to track a custom evaluation on an test dataset, model.stop_training is reset to False even if it was previously set to True by Early Stopping. And then the training does not stop. Current workaround is putting EarlyStopping as the last callback of the list.

Check this colab with the full example: https://colab.research.google.com/drive/1lw943Ggwkp_wvGxVX-5XqaEJ5qXmrHlA

This patch fixes this by checking first if stop_training exists, and only if it doesn't, it initializes it to False. 
Fixes #37587

### CODE REVIEW

**Comment by mihaimaruseac** [comment]
Please open against master branch, not against r2.1

**Comment by mihaimaruseac** [comment]
You will have to create a new one, unfortunately it cannot be automatically be forwarded to be on top of master

**Comment by eduardofv** [comment]
OK, I'll do... but master does not have yet `keras/engine/training_v2.py`. Updating only `callbacks.py` would be an incomplete solution. How should I proceed? (I'm new to this)

**Comment by eduardofv** [comment]
BTW, Why could it be that `training_v2.py` is not on the master branch? I use tf from docker image tag latest-gpu-py3-jupyter and tf.__version__ is 2.1.0. Is master still in 1.15?

**Comment by mihaimaruseac** [comment]
Whenever we do a new release, we cut the branch from master. So, for example, for the upcoming 2.2 release, we cut `r2.2` from master. Then on the branch we only get cherrypicks needed to unblock the release and a few commits to change version numbers and release notes and that is all.

If we do a patch release (within support window, 1 year usually, 3 years for long term support releases), then we accept other patches. Note that we usually do patch releases only for security issues.

In this case, since the issue this fixes is also raised by you, first thing to do would be to test if it also reproduces in master. If it does, then the PR should be opened against master (and hopefully merged as a cherry-pick before the final releases for 2.2).

If it does not reproduce in master, then the suggestion is to wait for the 2.2 release, which should have the issue fixed (if you can, please check that the issue is not happening in `tensorflow==2.2.0-rc0` which was released last week).

If you still want to fix this for 2.1 (and only for 2.1; it is fixed in master and 2.2 already), then we can reopen this PR but it will not get merged until we get a requirement to do a patch release against 2.1 (for security purposes mostly). Since 2.1 is LTS, this might mean this PR could stay open for 3 years before merging. If we don't do any patch release during this time, we will close the PR without merging.

The missing file on master could mean that the code got moved around (we are doing some keras refactorings at the time). The git history should tell you what happened with the file.

To summarize, I would:
1. test the issue against `tensorflow==2.2.0-rc0`. If it is solved, nothing to do
2. test the issue against master. Either build from source or install `tf-nightly`. If it is solved, but 1. above failed then we should do a cherry-pick of the fixing commit. If it fails, then we should recreate this against master.
3. decide if you want to patch 2.1, waiting until a patch release process starts. If yes, then I will reopen the PR (if master works) or I would suggest fixing master and then doing a cherry-pick PR instead (as that keeps a pointer in git history)

**Comment by eduardofv** [comment]
Excelent! It's clear and understood. I'll check if patch is still necesary and resubmit if needed. Thank you very much for your explanation!

**Comment by eduardofv** [comment]
Can confirm the issue is solved on v2.2. Thanks for your thorough explanation, it was very useful to me. Hope to be able to contribute in the future!
